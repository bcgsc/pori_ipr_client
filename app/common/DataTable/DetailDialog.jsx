import React from 'react';

function DetailDialog() {

}

export default DetailDialog;
