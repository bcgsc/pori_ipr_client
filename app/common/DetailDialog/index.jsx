import React from 'react';
import PropTypes from 'prop-types';
import Dialog from '@material-ui/core/Dialog';
import DialogTitle from '@material-ui/core/DialogTitle';
import DialogContent from '@material-ui/core/DialogContent';
import Typography from '@material-ui/core/Typography';
import Divider from '@material-ui/core/Divider';

import './index.scss';

/**
 * @param {object} props props
 * @param {func} props.onClose callback function to run on close
 * @param {object} props.selectedRow row data for selected row
 * @param {bool} props.open is the dialog open
 * @param {array} props.arrayColumns list of columns containing array data
 * @return {*} jsx
 */
function DetailDialog(props) {
  const {
    onClose,
    selectedRow,
    open,
    arrayColumns,
  } = props;

  const handleClose = (value) => {
    onClose(value);
  };

  return (
    <Dialog onClose={handleClose} open={open}>
      <DialogTitle>
        Detailed View
      </DialogTitle>
      <DialogContent>
        {Object.entries(selectedRow).map(([key, value], index) => (
          <React.Fragment key={key}>
            {index > 0 && <Divider />}
            <Typography className="detail-dialog__row">
              {key}
              {': '}
              {arrayColumns.includes(key) && (
                <>
                  {[...value].sort().map((val, i) => (
                    <React.Fragment key={val}>
                      {val.replace('#', '')}
                      {i < value.size - 1 && ', '}
                    </React.Fragment>
                  ))}
                </>
              )}
              {!arrayColumns.includes(key) && (
                <>
                  {value}
                </>
              )}
            </Typography>
          </React.Fragment>
        ))}
      </DialogContent>
    </Dialog>
  );
}

DetailDialog.propTypes = {
  onClose: PropTypes.func.isRequired,
  selectedRow: PropTypes.objectOf(PropTypes.any).isRequired,
  open: PropTypes.bool.isRequired,
  arrayColumns: PropTypes.arrayOf(PropTypes.string),
};

DetailDialog.defaultProps = {
  arrayColumns: [],
};

export default DetailDialog;
