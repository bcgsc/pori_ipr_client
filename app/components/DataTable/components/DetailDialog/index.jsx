import React from 'react';
import PropTypes from 'prop-types';
import Dialog from '@material-ui/core/Dialog';
import DialogTitle from '@material-ui/core/DialogTitle';
import DialogContent from '@material-ui/core/DialogContent';
import Typography from '@material-ui/core/Typography';
import Divider from '@material-ui/core/Divider';

import './index.scss';

const { compare } = new Intl.Collator(undefined, { numeric: true, sensitivity: 'base' });

/**
 * @param {object} props props
 * @param {func} props.onClose callback function to run on close
 * @param {object} props.selectedRow row data for selected row
 * @param {bool} props.open is the dialog open
 * @param {object} props.columnMapping mapping object for displayed col names
 * @return {*} jsx
 */
function DetailDialog(props) {
  const {
    onClose,
    selectedRow,
    open,
    columnMapping,
  } = props;

  const handleClose = (value) => {
    onClose(value);
  };

  const printRow = rows => Object.entries(rows)
    .sort(compare)
    .map(([key, value], index, rowEntries) => {
      if (key === 'ident' || value === null) {
        return null;
      }

      if (Array.isArray(value)) {
        return (
          <React.Fragment key={key}>
            <div className="detail-dialog__row">
              <Typography variant="subtitle2" display="inline">
                {`${columnMapping[key] || key}:  [`}
              </Typography>
            </div>
            {value.length ? (
              <>
                <div className="detail-dialog__inner-row">
                  {value.map(((arrVal, innerIndex) => (
                    // eslint-disable-next-line react/no-array-index-key
                    <React.Fragment key={innerIndex}>
                      {typeof arrVal === 'object' && arrVal !== null
                        ? (
                          <>
                            <Divider />
                            {printRow(arrVal)}
                          </>
                        )
                        : (
                          <Typography variant="body2">
                            {arrVal || 'null'}
                          </Typography>
                        )}
                    </React.Fragment>
                  )))}
                </div>
              </>
            ) : null}
            <div className="detail-dialog__row">
              <Typography variant="subtitle2" display="inline">
                {']'}
              </Typography>
            </div>
            <Divider />
          </React.Fragment>
        );
      }

      if (typeof value === 'object' && value !== null) {
        return (
          <React.Fragment key={key}>
            <div className="detail-dialog__row">
              <Typography variant="subtitle2">
                {`${columnMapping[key] || key}:  {`}
              </Typography>
            </div>
            <div className="detail-dialog__inner-row">
              {printRow(value)}
            </div>
            <div className="detail-dialog__row">
              <Typography variant="subtitle2">
                {'}'}
              </Typography>
            </div>
            <Divider />
          </React.Fragment>
        );
      }

      return (
        <React.Fragment key={key}>
          <div className="detail-dialog__row">
            <Typography variant="subtitle2" display="inline">
              {`${columnMapping[key] || key}: `}
            </Typography>
            <Typography variant="body2" display="inline">
              {`${value}` || 'null'}
            </Typography>
          </div>
          {index !== rowEntries.length - 1 && <Divider />}
        </React.Fragment>
      );
    });


  return (
    <Dialog onClose={handleClose} open={open}>
      <DialogTitle>
        Detailed View
      </DialogTitle>
      <DialogContent>
        {printRow(selectedRow)}
      </DialogContent>
    </Dialog>
  );
}

DetailDialog.propTypes = {
  onClose: PropTypes.func.isRequired,
  selectedRow: PropTypes.objectOf(PropTypes.any).isRequired,
  open: PropTypes.bool.isRequired,
  columnMapping: PropTypes.objectOf(PropTypes.string),
};

DetailDialog.defaultProps = {
  columnMapping: {},
};

export default DetailDialog;
