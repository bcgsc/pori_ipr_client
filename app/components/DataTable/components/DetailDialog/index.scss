.detail-dialog {
  &__row {
    padding: 8px 4px;
  }

  &__inner-row {
    padding: 0 12px;
  }
}
