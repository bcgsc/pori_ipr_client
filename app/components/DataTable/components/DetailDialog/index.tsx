import React from 'react';
import Dialog from '@mui/material/Dialog';
import DialogTitle from '@mui/material/DialogTitle';
import DialogContent from '@mui/material/DialogContent';
import Typography from '@mui/material/Typography';
import Divider from '@mui/material/Divider';

import './index.scss';

const { compare } = new Intl.Collator(undefined, { numeric: true, sensitivity: 'base' });

type DetailDialogProps = {
  onClose: (value: string) => void,
  selectedRow,
  isOpen: boolean,
  columnMapping?: Record<string, string>,
};

/**
 * @param {object} props props
 * @param {func} props.onClose callback function to run on close
 * @param {object} props.selectedRow row data for selected row
 * @param {bool} props.isOpen is the dialog open
 * @param {object} props.columnMapping mapping object for displayed col names
 * @return {*} jsx
 */
const DetailDialog = ({
  onClose,
  selectedRow,
  isOpen,
  columnMapping = {},
}: DetailDialogProps): JSX.Element => {
  const renderRow = (rows) => Object.entries(rows)
    .sort(([a], [b]) => compare(a, b))
    .map(([key, value], index, rowEntries) => {
      if (['ident', 'svg', 'svgTitle', 'image'].includes(key) || value === null) {
        return null;
      }

      if (Array.isArray(value)) {
        return (
          <React.Fragment key={key}>
            <div className="detail-dialog__row">
              <Typography variant="subtitle2" display="inline">
                {`${columnMapping[key] || key}:  [`}
              </Typography>
            </div>
            {value.length ? (
              <div className="detail-dialog__inner-row">
                {value.map(((arrVal, innerIndex) => (
                  // eslint-disable-next-line react/no-array-index-key
                  <React.Fragment key={innerIndex}>
                    {typeof arrVal === 'object' && arrVal !== null
                      ? (
                        <>
                          <Divider />
                          {renderRow(arrVal)}
                        </>
                      )
                      : (
                        <Typography variant="body2">
                          {arrVal || 'null'}
                        </Typography>
                      )}
                  </React.Fragment>
                )))}
              </div>
            ) : null}
            <div className="detail-dialog__row">
              <Typography variant="subtitle2" display="inline">
                ]
              </Typography>
            </div>
            <Divider />
          </React.Fragment>
        );
      }

      if (typeof value === 'object' && value !== null) {
        return (
          <React.Fragment key={key}>
            <div className="detail-dialog__row">
              <Typography variant="subtitle2">
                {`${columnMapping[key] || key}:  {`}
              </Typography>
            </div>
            <div className="detail-dialog__inner-row">
              {renderRow(value)}
            </div>
            <div className="detail-dialog__row">
              <Typography variant="subtitle2">
                {'}'}
              </Typography>
            </div>
            <Divider />
          </React.Fragment>
        );
      }

      return (
        <React.Fragment key={key}>
          <div className="detail-dialog__row">
            <Typography variant="subtitle2" display="inline">
              {`${columnMapping[key] || key}: `}
            </Typography>
            <Typography variant="body2" display="inline">
              {`${value}` || 'null'}
            </Typography>
          </div>
          {index !== rowEntries.length - 1 && <Divider />}
        </React.Fragment>
      );
    });

  return (
    <Dialog onClose={onClose} open={isOpen}>
      <DialogTitle>
        Detailed View
      </DialogTitle>
      <DialogContent>
        {renderRow(selectedRow)}
      </DialogContent>
    </Dialog>
  );
};

export default DetailDialog;
