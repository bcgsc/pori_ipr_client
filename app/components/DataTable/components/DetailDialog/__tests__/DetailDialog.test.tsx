import React from 'react';
import { screen, render, fireEvent } from '@testing-library/react';

import DetailDialog from '..';

const mockData = {
  detectedIn: 'DNA',
  gene: {
    name: 'EGFR',
    isImportant: true,
    innerObject: {
      innerInnerValue: 'double nested',
    },
  },
};

const mockArrayData = {
  arrayVal: ['test1', 'test2'],
};

const mockFilteredData = {
  ident: '640aff03-402bdd42-dffa-623b',
  svg: '<svg><g>image here</g></svg>',
  svgTitle: 'test SVG',
  image: 'test image',
};

const mockColumnMapping = {
  detectedIn: 'Detected In',
};

describe('DetailDialog', () => {
  test('It matches the snapshot', () => {
    const { asFragment } = render(
      <DetailDialog
        isOpen
        selectedRow={mockData}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('Keys and values are all shown', () => {
    render(
      <DetailDialog
        isOpen
        selectedRow={mockData}
      />,
    );
    Object.entries(mockData).forEach(([key, val]) => {
      expect(screen.getByText(key, { exact: false })).toBeInTheDocument();
      if (typeof val === 'object') {
        // naive implementation to test recursion in the component
        Object.entries(val).forEach(([innerKey, innerVal]) => {
          expect(screen.getByText(innerKey, { exact: false })).toBeInTheDocument();
          if (typeof innerVal === 'object') {
            Object.entries(innerVal).forEach(([doubleInnerKey, doubleInnerVal]) => {
              expect(screen.getByText(doubleInnerKey, { exact: false })).toBeInTheDocument();
              expect(screen.getByText(doubleInnerVal)).toBeInTheDocument();
            });
          } else {
            expect(screen.getByText(`${innerVal}`)).toBeInTheDocument();
          }
        });
      } else {
        expect(screen.getByText(val)).toBeInTheDocument();
      }
    });
  });

  test('Array values are shown', () => {
    render(
      <DetailDialog
        isOpen
        selectedRow={mockArrayData}
      />,
    );

    Object.entries(mockArrayData).forEach(([key, val]) => {
      expect(screen.getByText(key, { exact: false })).toBeInTheDocument();
      val.forEach((v) => {
        expect(screen.getByText(v, { exact: false })).toBeInTheDocument();
      });
    });
  });

  test('The column mapping changes the displayed key', async () => {
    render(
      <DetailDialog
        columnMapping={mockColumnMapping}
        isOpen
        selectedRow={mockData}
      />,
    );

    expect(await screen.findByText(`${mockColumnMapping.detectedIn}:`)).toBeInTheDocument();
    expect(screen.queryByText('detectedIn', { exact: false })).toBeNull();
  });

  test('Certain fields are filtered from being shown', () => {
    render(
      <DetailDialog
        isOpen
        selectedRow={mockFilteredData}
      />,
    );

    Object.entries(mockFilteredData).forEach(([key, val]) => {
      expect(screen.queryByText(key)).toBeNull();
      expect(screen.queryByText(val)).toBeNull();
    });
  });
});
