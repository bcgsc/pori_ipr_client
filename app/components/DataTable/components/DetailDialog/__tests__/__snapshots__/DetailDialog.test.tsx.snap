// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailDialog It matches the snapshot 1`] = `<DocumentFragment />`;
